src/IO/HTTPCommon.cpp

```diff
@@ -51,6 +51,58 @@ namespace
         session.setKeepAliveTimeout(timeouts.http_keep_alive_timeout);
     }
 
+    template <typename Session>
+    class HTTPSessionAdapter : public Session
+    {
+    public:
+        HTTPSessionAdapter(const std::string & host, UInt16 port) :
+            Session(host,port)
+        {
+
+        }
+        ~HTTPSessionAdapter() override = default;
+
+    protected:
+       void reconnect() override
+       {
+            // First of all will try to establish connection with previous addr.
+            if (!Session::getResolvedHost().empty())
+            {
+                try
+                {
+                    Session::reconnect();
+                    return;
+                }
+                catch (...)
+                {
+                    LOG_TRACE((&Poco::Logger::get("HTTPSessionAdapter")), "Last ip ({})  is unreachable for {}:{}. Will try another resolved address.",
+                                                                           Session::getResolvedHost(), Session::getHost(), Session::getPort());
+                }
+            }
+
+            const auto& endpoinds = DNSResolver::instance().resolveHostAll(Session::getHost());
+
+            for (auto it = endpoinds.begin();;)
+            {
+                try {
+                    Session::setResolvedHost(it->toString());
+                    Session::reconnect();
+                    LOG_TRACE((&Poco::Logger::get("HTTPSessionAdapter")), "Created HTTP(S) session with {}:{} ({}:{})", Session::getHost(), Session::getPort(),
+                                                                           it->toString(), Session::getPort());
+                    break;
+                }
+                catch (...)
+                {
+                    Session::close();
```

Comment:
It is probably worth to `setResolvedHost("")`  here or do this in `HttpClientSession::reset()` and call it here